### DISCUSSION TITLE
Multispectral Imagery and Channel FX

### DISCUSSION THREAD

**Discussion by urban-warrior** [ImageMagick/ImageMagick#6156] [General]
@fmw42 asks "Can't you use -channel-fx to build the multispectral image from multiple grayscale images?  That is take an RGB or CMYK image and add other grayscale images to it in the meta channels using -channel-fx"

**Comment by urban-warrior** [comment]
You could, for example, copy a gray channel into a new meta-channel.  Then repeat to add a second meta channel.  For the first example, try:

```
magick gray.pgm -channel-fx "0=>meta" -verbose meta.tif
```

**Reply by fmw42** [reply]
1) Can you show the command to add a grayscale image to the meta channel of an rgb image?

2) Can one add multiple grayscale images at the same time to an rgb image with -channel-fx?

**Reply by urban-warrior** [reply]
For 1), use the same same command above but use an RGB image instead.  For 2) use a comma, e.g., `-channel-fx "0=>meta, 1=>meta1"`.

**Reply by fmw42** [reply]
I want to add grayscale images to an existing RGB image.

So for 1) do you mean

`magick rgb.png gray.pgm -channel-fx "0=>meta" -verbose meta.tif`

or should it not be

`magick rgb.png gray.pgm -channel-fx '| gray=>meta' meta.tif`

and for 2)

`magick rgb.png gray1.pgm gray2.pgm -channel-fx "0=>meta, 1=>meta1" -verbose meta.tif`

and should this not be using | also and if so, how?

`magick rgb.png gray1.pgm gray2.pgm -channel-fx "|0=>meta, |1=>meta1" -verbose meta.tif`

**Reply by urban-warrior** [reply]
Explore the channel FX option and let us know what works and what doesn't.  If something doesn't work and you think it should, post here and we'll investigate and if it turns out its a bug, we'll add a patch to resolve the issue.

**Reply by fmw42** [reply]
It would be nice to see those examples above in correct form from you. I am not sure I understand all the nuances of -channel-fx. That would point me in the right direction to explore them further.

**Comment by fmw42** [comment]
1) This works to put one grayscale image into the meta channel of the RGB image. It needs the | symbol.


Inputs:

![lena](https://user-images.githubusercontent.com/13458218/224573840-6a1fde50-38f5-4de0-8653-3179f4e503d5.png)

![mandril3](https://user-images.githubusercontent.com/13458218/224573844-21a78f75-8199-484d-b3ef-beeaff9dab3e.jpg)

![zelda1](https://user-images.githubusercontent.com/13458218/224573849-94916837-843f-4dc8-a50e-c841ec6cc1e0.jpg)


```
magick lena.png \( mandril3.jpg -colorspace gray \) -channel-fx '| gray=>meta' meta.tif
magick identify -verbose meta.tif

   Red:
      min: 53  (0.207843)
      max: 255 (1)
      mean: 180.23 (0.706786)
      median: 196 (0.768627)
      standard deviation: 48.8793 (0.191684)
      kurtosis: -0.792362
      skewness: -0.695346
      entropy: 0.947403
    Green:
      min: 0  (0)
      max: 253 (0.992157)
      mean: 99.0252 (0.388334)
      median: 97 (0.380392)
      standard deviation: 52.5374 (0.206029)
      kurtosis: -0.738533
      skewness: 0.231172
      entropy: 0.961838
    Blue:
      min: 43  (0.168627)
      max: 216 (0.847059)
      mean: 105.384 (0.413273)
      median: 101 (0.396078)
      standard deviation: 33.4045 (0.130998)
      kurtosis: -0.141787
      skewness: 0.682631
      entropy: 0.932753
    Meta channel[0]:
      min: 3  (0.0117647)
      max: 220 (0.862745)
      mean: 129.588 (0.508187)
      median: 129 (0.505882)
      standard deviation: 39.6973 (0.155676)
      kurtosis: -0.596356
      skewness: -0.206788
      entropy: 0.938578

```

**But I see the effects of the meta channel in the result as if it were in the alpha channel.**

[meta.tif.zip](https://github.com/ImageMagick/ImageMagick/files/10952087/meta.tif.zip)


---------------------------------


2) But the following seems to put the same image (the first grayscale image) into both meta channels. 

**How do I get it to put two different grayscale images into the RGB image's meta channels.**

```
magick lena.png \( mandril3.jpg -colorspace gray \) \( zelda1.jpg -colorspace gray \) -channel-fx '| gray=>meta, | gray=>meta1' meta.tif
magick identify -verbose meta.tif

   Red: 8-bit
    Green: 8-bit
    Blue: 8-bit
    Meta channel[0]: 8-bit
    Meta channel[1]: 8-bit
  Channel statistics:
    Pixels: 65536
    Red:
      min: 53  (0.207843)
      max: 255 (1)
      mean: 180.23 (0.706786)
      median: 196 (0.768627)
      standard deviation: 48.8793 (0.191684)
      kurtosis: -0.792362
      skewness: -0.695346
      entropy: 0.947403
    Green:
      min: 0  (0)
      max: 253 (0.992157)
      mean: 99.0252 (0.388334)
      median: 97 (0.380392)
      standard deviation: 52.5374 (0.206029)
      kurtosis: -0.738533
      skewness: 0.231172
      entropy: 0.961838
    Blue:
      min: 43  (0.168627)
      max: 216 (0.847059)
      mean: 105.384 (0.413273)
      median: 101 (0.396078)
      standard deviation: 33.4045 (0.130998)
      kurtosis: -0.141787
      skewness: 0.682631
      entropy: 0.932753
    Meta channel[0]:
      min: 3  (0.0117647)
      max: 220 (0.862745)
      mean: 129.588 (0.508187)
      median: 129 (0.505882)
      standard deviation: 39.6973 (0.155676)
      kurtosis: -0.596356
      skewness: -0.206788
      entropy: 0.938578
    Meta channel[1]:
      min: 3  (0.0117647)
      max: 220 (0.862745)
      mean: 129.588 (0.508187)
      median: 129 (0.505882)
      standard deviation: 39.6973 (0.155676)
      kurtosis: -0.596356
      skewness: -0.206788
      entropy: 0.938578

```

**Reply by urban-warrior** [reply]
Try using parens and more than one `-channel-fx` option, one for each gray image and meta channel.

**Reply by fmw42** [reply]
OK, this seems to work to put two grayscale images into two meta channels.

```
magick lena.png \( mandril3.jpg -colorspace gray \) -channel-fx '| gray=>meta' \( zelda1.jpg -colorspace gray \) -channel-fx '| gray=>meta1' meta.tif
magick identify -verbose meta.tif

   Red:
      min: 53  (0.207843)
      max: 255 (1)
      mean: 180.23 (0.706786)
      median: 196 (0.768627)
      standard deviation: 48.8793 (0.191684)
      kurtosis: -0.792362
      skewness: -0.695346
      entropy: 0.947403
    Green:
      min: 0  (0)
      max: 253 (0.992157)
      mean: 99.0252 (0.388334)
      median: 97 (0.380392)
      standard deviation: 52.5374 (0.206029)
      kurtosis: -0.738533
      skewness: 0.231172
      entropy: 0.961838
    Blue:
      min: 43  (0.168627)
      max: 216 (0.847059)
      mean: 105.384 (0.413273)
      median: 101 (0.396078)
      standard deviation: 33.4045 (0.130998)
      kurtosis: -0.141787
      skewness: 0.682631
      entropy: 0.932753
    Meta channel[0]:

      min: 3  (0.0117647)
      max: 220 (0.862745)
      mean: 129.588 (0.508187)
      median: 129 (0.505882)
      standard deviation: 39.6973 (0.155676)
      kurtosis: -0.596356
      skewness: -0.206788
      entropy: 0.938578
    Meta channel[1]:
      min: 3  (0.0117647)
      max: 255 (1)
      mean: 141.048 (0.553131)
      median: 146 (0.572549)
      standard deviation: 58.3115 (0.228672)
      kurtosis: -0.848452
      skewness: -0.207985
      entropy: 0.975697


```


**But what about my issue with the result from 1) above showing the meta channel in the alpha channel.**

**Reply by urban-warrior** [reply]
Not sure what the issue is.  This shows there is no alpha channel:

```
$ tiffinfo meta.tif
=== TIFF directory 0 ===
TIFF Directory at offset 0x40008 (262152)
  Image Width: 256 Image Length: 256
  Bits/Sample: 8
  Compression Scheme: None
  Photometric Interpretation: RGB color
  Extra Samples: 1<unspecified>
  FillOrder: msb-to-lsb
  Orientation: row 0 top, col 0 lhs
  Samples/Pixel: 4
  Rows/Strip: 256
  Planar Configuration: single image plane
  Page Number: 0-1
  White Point: 0.3127-0.329
  PrimaryChromaticities: 0.640000,0.330000,0.300000,0.600000,0.150000,0.060000
```

It's an RGB image with a meta channel, as expected.  If it was alpha, it would show Extra Samples as `unassoc-alpha` rather than `unspecified`.
